src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2607,7 +2716,21 @@ protected virtual void MarkInterfaceImplementation (InterfaceImplementation ifac
 
 		bool HasManuallyTrackedDependency (MethodBody methodBody)
 		{
-			return PreserveDependencyLookupStep.HasPreserveDependencyAttribute (methodBody.Method);
+			var method = methodBody.Method;
+
+			if (!method.HasCustomAttributes)
+				return false;
+
+			// avoid allocating linker attributes for members that don't need them
+			foreach (var ca in method.CustomAttributes) {
+				if (LinkerAttributesInformation.IsAttribute<DynamicDependencyAttribute> (ca.AttributeType))
+					return true;
+#if !FEATURE_ILLINK
+				if (DynamicDependencyLookupStep.IsPreserveDependencyAttribute (ca.AttributeType))
+					return true;
```

Comment:
I would very much like to avoid doing this. See https://github.com/mono/linker/issues/1140.

Since we have the suppression mechanism maybe that is the better solution for this problem. Also in most cases where we use it, the code will not generate warnings, we use the attribute as basically just a workaround for the assembly loading issue.